ENIGMAsystem/SHELL/Universal_System/Extensions/Steamworks/game_client/c_leaderboards.cpp

```diff
@@ -0,0 +1,95 @@
+
+  if (steam_api_call != 0) {
+    c_leaderboards::m_callResultFindLeaderboard.Set(steam_api_call, this, &c_leaderboards::on_find_leaderboard);
+    c_leaderboards::loading_ = true;
+  }
+}
+
+void c_leaderboards::on_find_leaderboard(LeaderboardFindResult_t* pFindLeaderboardResult, bool bIOFailure) {
+  c_leaderboards::loading_ = false;
+
+  if (!pFindLeaderboardResult->m_bLeaderboardFound || bIOFailure) return;
+
+  c_leaderboards::current_leaderboard_ = pFindLeaderboardResult->m_hSteamLeaderboard;
+
+  DEBUG_MESSAGE("Calling FindOrCreateLeaderboard succeeded.", M_INFO);
+}
+
+bool c_leaderboards::upload_score(const int score,
+                                  const ELeaderboardUploadScoreMethod leaderboard_upload_score_method) {
+  // while (c_leaderboards::loading_)
+  //   ;  // Wait for the callback of FindOrCreateLeaderboard to be invoked
+
+  if (NULL == c_leaderboards::current_leaderboard_) return false;
+
+  SteamAPICall_t steam_api_call = SteamUserStats()->UploadLeaderboardScore(
+      c_leaderboards::current_leaderboard_, leaderboard_upload_score_method, score, NULL, 0);
+  c_leaderboards::m_SteamCallResultUploadScore.Set(steam_api_call, this, &c_leaderboards::on_upload_score);
+
+  return true;
+}
+
+void c_leaderboards::on_upload_score(LeaderboardScoreUploaded_t* pScoreUploadedResult, bool bIOFailure) {
+  if (!pScoreUploadedResult->m_bSuccess || bIOFailure) {
+    DEBUG_MESSAGE("Failed to upload score to leaderboard.", M_ERROR);
+  }
+}
+
+bool c_leaderboards::download_scores(const ELeaderboardDataRequest leaderboard_data_request, const int range_start,
+                                     const int range_end) {
+  // while (c_leaderboards::loading_)
+  //   ;  // Wait for the callback of FindOrCreateLeaderboard to be invoked
+
+  if (NULL == c_leaderboards::current_leaderboard_) return false;
+
+  SteamAPICall_t steam_api_call = SteamUserStats()->DownloadLeaderboardEntries(
+      c_leaderboards::current_leaderboard_, leaderboard_data_request, range_start, range_end);
+  c_leaderboards::m_SteamCallResultDownloadScores.Set(steam_api_call, this, &c_leaderboards::on_download_scores);
+
+  return true;
+}
+
+void c_leaderboards::on_download_scores(LeaderboardScoresDownloaded_t* pLeaderboardScoresDownloaded, bool bIOFailure) {
+  if (bIOFailure) {
+    DEBUG_MESSAGE("Failed to download scores from leaderboard.", M_ERROR);
+    return;
+  }
+
+  DEBUG_MESSAGE("Downloaded scores from leaderboard.", M_INFO);
+
+  // enigma::signal(&enigma::asyncsteamworks_mutex);
+
+  // int map = enigma_user::ds_map_create();
+
+  // enigma_user::ds_map_add(map, "event_type", "leaderboard_downloaded_scores");
+  // enigma_user::ds_map_add(map, "leaderboard", c_leaderboards::current_leaderboard_);
+
+  enigma::posted_async_events.push(std::map<std::string, variant>{
+      {"event_type", "leaderboard_downloaded_scores"}, {"leaderboard", c_leaderboards::current_leaderboard_}});
```

Comment:
@RobertBColton 

I testing this pushing and I'm sure that it's pushing to the queue perfectly, so the problem must be from the fire function.

Reply:
We talked about this, it was due to serialization issues with JSON and mutual exclusivity.